private/bufpkg/bufwasm/bufwasm.go

```diff
@@ -0,0 +1,255 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufwasm
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"testing/fstest"
+
+	wasmpluginv1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/wasmplugin/v1"
+	"github.com/bufbuild/buf/private/pkg/protoencoding"
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+	gojs "github.com/tetratelabs/wazero/imports/go"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
+	"github.com/tetratelabs/wazero/sys"
+	"go.uber.org/multierr"
+	"google.golang.org/protobuf/encoding/protowire"
+	"google.golang.org/protobuf/proto"
+)
+
+// CustomSectionName is the name of the custom wasm section we look into for buf
+// extensions.
+const CustomSectionName = ".bufplugin"
+
+// https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#sections%E2%91%A0
+const customSectionID = 0
+
+// EncodeBufSection encodes the pluginmetadata message as a custom wasm section.
+// The resulting bytes can be appended to any valid wasm file to add the new
+// section to that file.
+func EncodeBufSection(metadata *wasmpluginv1.Metadata) ([]byte, error) {
+	metadataBinary, err := protoencoding.NewWireMarshaler().Marshal(metadata)
+	if err != nil {
+		return nil, err
+	}
+	// Abusing the protowire package because the wasm file format is similar.
+	return protowire.AppendBytes(
+		[]byte{customSectionID},
+		append(
+			protowire.AppendString(
+				nil,
+				CustomSectionName,
+			),
+			metadataBinary...,
+		),
+	), nil
+}
+
+// CompiledPlugin is the compiled representation of loading wasm bytes.
+type CompiledPlugin struct {
+	Module wazero.CompiledModule
+
+	// Metadata parsed from custom sections of the wasm file. May be nil if
+	// no buf specific sections were found.
+	Metadata *wasmpluginv1.Metadata
+}
+
+func (c *CompiledPlugin) ABI() wasmpluginv1.WasmABI {
+	if c.Metadata.GetAbi() != wasmpluginv1.WasmABI_WASM_ABI_UNSPECIFIED {
+		return c.Metadata.GetAbi()
+	}
+	exportedFuncs := c.Module.ExportedFunctions()
+	if _, ok := exportedFuncs["_start"]; ok {
+		return wasmpluginv1.WasmABI_WASM_ABI_WASI_SNAPSHOT_PREVIEW1
+	} else if _, ok := exportedFuncs["run"]; ok {
+		return wasmpluginv1.WasmABI_WASM_ABI_GOJS
+	}
+	return wasmpluginv1.WasmABI_WASM_ABI_UNSPECIFIED
+}
+
+// PluginExecutorOption are options
+type PluginExecutorOption func(*PluginExecutor)
+
+// WithMemoryLimitPages provides a custom per memory limit for a plugin
+// executor. The default is 8192 pages for 512MB.
+func WithMemoryLimitPages(memoryLimitPages uint32) PluginExecutorOption {
+	return func(pluginExecutor *PluginExecutor) {
+		pluginExecutor.runtimeConfig = pluginExecutor.runtimeConfig.WithMemoryLimitPages(memoryLimitPages)
```

Comment:
We might consider allowing this to be specified by `buf generate` as a CLI option.